hardhat: fix docs according to expected kusama launch (#563)
polkadot-developers/polkadot-docs: develop/smart-contracts/dev-environments/hardhat.md

- [Node.js](https://nodejs.org/){target=\_blank} (v16.0.0 or later) and npm installed
- Basic understanding of Solidity programming
- Some WND test tokens to cover transaction fees (easily obtainable from the [Polkadot faucet](https://faucet.polkadot.io/westend?parachain=1000){target=\_blank}). To learn how to get test tokens, check out the [Test Tokens](/develop/smart-contracts/connect-to-asset-hub/#test-tokens){target=\_blank} section
- [MetaMask](https://metamask.io/){target=\_blank} installed and connected to [Westend Hub](https://chainlist.org/chain/420420421){target=\_blank}. For more detailed instructions on connecting your wallet, see the [Connect Your Wallet](/develop/smart-contracts/connect-to-asset-hub/#connect-your-wallet){target=\_blank} section

> ## Setting Up Hardhat

Reviewer: This whole section needs to be updated, now that we have `npx hardhat-polkadot init`.

Reviewer: If that's the goal, it makes sense :+1:

    npm install --save-dev hardhat @nomicfoundation/hardhat-toolbox
    ```

    To interact with Polkadot Hub, Hardhat requires a plugin to compile contracts to PolkaVM bytecode and the and to spawn a local node compatible with PolkaVM.

    ```bash

>     npm install --save-dev @parity/hardhat-polkadot

Author: We are releasing an umbrella package to 1) simplify onboarding 2) simplify adding more packages without the user having to install them

>     npm install --save-dev @parity/hardhat-polkadot

Author: What do you think about this move?

>     npm install --save-dev @parity/hardhat-polkadot

Reviewer 2: Yes, having only one package will be much better for devs, I fully agree with that. However, I was suggesting that we need to implement this change only after the package has been deployed and we confirm that the command `npm install @parity/hardhat-polkadot` actually works

>     npm install --save-dev @parity/hardhat-polkadot

Author: Absolutely. This PR should only be merged after the package is available

>     npm install --save-dev @parity/hardhat-polkadot

Reviewer 2: ah okay, I'd put this PR as a draft PR then, and it will get unblock once that package is available :)

5. Update your Hardhat configuration file (`hardhat.config.js`) to include the plugins:

    ```javascript title="hardhat.config.js" hl_lines="8-8"
    --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:0:10'
    --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:13:14'
      // Additional configuration will be added later
    --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:57:57'
    ```

## Compiling Your Contract

The plugin will compile your Solidity contracts for Solidity versions `0.8.0` and higher to be PolkaVM compatible. When compiling your contract, there are two ways to configure your compilation process:

- **Npm compiler** - uses library [@parity/revive](https://www.npmjs.com/package/@parity/revive){target=\_blank} for simplicity and ease of use

> - **Binary compiler** - uses the resolc binary directly for more control and configuration options

Reviewer: ```suggestion
- **Binary compiler** - uses your local `resolc` binary directly for more control and configuration options
```

Author: Thanks!

## Compiling Your Contract

The plugin will compile your Solidity contracts for Solidity versions `0.8.0` and higher to be PolkaVM compatible. When compiling your contract, there are two ways to configure your compilation process:

> - **Npm compiler** - uses library [@parity/revive](https://www.npmjs.com/package/@parity/revive){target=\_blank} for simplicity and ease of use

Author: > Can the @parity/revive be integrated into the hardhat workflow

@parity/revive is already integrated right? Users don't have to install it. 

> Would it be possible to add a complete example of how to use npm

Is [npm.config.ts](https://github.com/paritytech/hardhat-polkadot/blob/main/examples/npm.config.ts) enough? I am proposing we add full project examples and not just the hardhat.config.ts? What do you think? In addition to this, do you think users will use binary configuration? What if they don't even add this "resolc" key to config? We just compile the thing if the package is imported and if someone wants to debug the compiler they can configure the "resolc" with the binary. Looking forward to your feedback

> - **Npm compiler** - uses library [@parity/revive](https://www.npmjs.com/package/@parity/revive){target=\_blank} for simplicity and ease of use

Reviewer 2: We do have some full project examples available here - https://github.com/polkadot-developers/polkavm-hardhat-examples

From my experience, the binary configuration was quite complex from a developer experience standpoint. I had to download the binary, switch the solc version on my machine to a compatible one, and then check out and compile a matching version of the Polkadot SDK to generate a compatible substrate-node and eth-rpc-adapter.

Now that there's an npm-based option, I think we should emphasize using that as much as possible.

That said, for this guide—which serves as a starting point for developers—I suggest we explain both approaches. We can highlight the npm.config.ts setup (https://github.com/paritytech/hardhat-polkadot/blob/main/examples/npm.config.ts) and pull that file directly from the Hardhat repo to avoid duplication. After covering the npm path, we can still include the binary configuration for projects that might require customizing the resolc binary, as you mentioned.

@BigTava, what do you think?

> - **Npm compiler** - uses library [@parity/revive](https://www.npmjs.com/package/@parity/revive){target=\_blank} for simplicity and ease of use

Author: > We do have some full project examples available here

Nice! Where will you feature these? I think it is key we have an examples folder too in our repo. They will also be part of our e2e test suite. Maybe you could just link to our folder so we reduce maintenance overhead. 

> what do you think

I think we shouldn't frame 2 options. We just mention there is a more advanced option to compile the contracts. The user shouldn't even need to configure the compiler. It just needs to import the package and it uses the npm as default. For now, let's keep as is and rename Remix -> Npm. But, if you agree, we would reduce the step "Compiling Your Contract" to just step 2 and 3 and briefly mention the binary configuration.

> - **Npm compiler** - uses library [@parity/revive](https://www.npmjs.com/package/@parity/revive){target=\_blank} for simplicity and ease of use

Reviewer 2: > Nice! Where will you feature these? I think it is key we have an examples folder too in our repo. They will also be part of our e2e test suite. Maybe you could just link to our folder so we reduce maintenance overhead.

I believe that for tutorials, we should keep them in the repository I shared previously since it would be easier to update those projects as the tutorial requires if we need to adjust something in the documentation. However, if you all add some projects and examples in the Hardhat Revive repository, we can also add references so that people can check different use cases as soon as those projects are pinned to a specific tag or release, ensuring that the documentation does not break even if the project gets updated.

> I think we shouldn't frame 2 options. We just mention there is a more advanced option to compile the contracts. The user shouldn't even need to configure the compiler. It just needs to import the package and it uses the npm as default. For now, let's keep as is and rename Remix -> Npm. But, if you agree, we would reduce the step "Compiling Your Contract" to just step 2 and 3 and briefly mention the binary configuration.

I like this, yeah I agree!

        --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:65:66'
        ```

>     For the binary configuration, check the [installation](https://github.com/paritytech/revive?tab=readme-ov-file#installation){target=\_blank} section of the `pallet-revive`.

Reviewer 2: Could you please roll back this change?

    ```

5. Update your Hardhat configuration file with network settings for the Polkadot network you want to target:

    === "Npm Configuration"

        ```javascript title="hardhat.config.js" hl_lines="37-41"
        --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:0:20'
          --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:34:55'
        --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:56:57'
        ```

    === "Binary Configuration"

        ```javascript title="hardhat.config.js" hl_lines="42-46"
        --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:0:10'
        --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:13:15'
          --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:22:33'
        --8<-- 'code/develop/smart-contracts/dev-environments/hardhat/hardhat.config.js:34:57'
        ```

5. Start a local node, if not running already:

    ```bash
    npx hardhat node
    ```

    This command will start a local PolkaVM node

6. Deploy your contract using Ignition:

    ```bash

>     npx hardhat ignition deploy ./ignition/modules/Lock.js --network westendHub

Reviewer: If we are running a local node in the previous step, we should use `--network localhost`. Or if we want to make a live deployment the step 5 doesn't make sense.

Author: Removed step 5

Run your interaction script:

```bash

> npx hardhat run scripts/interact.js --network polkadotHub

Reviewer 2: Mmm, I do understand your point, and I think @0xLucca worked in a PR to emphasize the usage of Polkadot Hub here - https://github.com/polkadot-developers/polkadot-docs/pull/572/files

But in this case, how would we differentiate if a user is trying to interact with Westend Asset Hub or polkadot asset hub? That's why I thought about using Westend Hub, as [it'd be changed here](https://github.com/polkadot-developers/polkadot-docs/pull/572/files#diff-9c331e9d30ae9a2f2f2bedbb90ac5390a32277b0414ae729830f8ef8ce05d2c3R18)

> npx hardhat run scripts/interact.js --network polkadotHub

Author: Exactly, the PR 572 is good. You're right, what about just calling it here in the docs polkadotTestnet?

> npx hardhat run scripts/interact.js --network polkadotHub

Reviewer 2: Totally, I agree on that point. I'm just a bit concerning about Paseo, because if, at some point, Paseo is also upgraded to support polkavm, the word `polkadotTestnet` would not be enough to cover both, westend and paseo. What do you think?

> npx hardhat run scripts/interact.js --network polkadotHub

Author: Good point. I will confirm that point. IMO developers only care, if testnet 1 is better than testnet 2 for some reason. If not, they just care it is a testnet

Run your interaction script:

```bash

> npx hardhat run scripts/interact.js --network westendHub

Reviewer: Same comment as above
